### PR TITLE
[Doc]  losses 

### PR DESCRIPTION
**Pull request by plyfager**

### CODE REVIEW

File: docs/en/advanced_guides/losses.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
+# Losses
+
+We provide two ways of defining loss.
+One is using default loss configuration, which has fixed loss components, and offer related loss arguments for users to adjust.
```
Comment (Z-Fran) [inline docs/en/advanced_guides/losses.md:4]: typo `offer`

File: docs/en/advanced_guides/losses.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+                self.discriminator,
+                curr_scale=self.curr_scale[0],
+                transition_weight=transition_weight))
+
+        loss_disc, log_vars_disc = self._get_disc_loss(data_dict_)
+
+        # ...
+```
+
+Here, the `_get_disc_loss` defined in [BaseGAN](https://github.com/open-mmlab/mmgeneration/tree/master/mmgen/models/gans/base_gan.py) will help to combine all kinds of losses automatically.
```
Comment (Z-Fran) [inline docs/en/advanced_guides/losses.md:144]: This link should be like [models](models.md) in line 31 this file

File: docs/en/advanced_guides/losses.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
+# Losses
+
+We provide two ways of defining loss.
```
Comment (zengyh1900) [inline docs/en/advanced_guides/losses.md:3]: ... two ways to design your own loss function.

File: docs/en/advanced_guides/losses.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+One is using default loss configuration, which has fixed loss components, and offer related loss arguments for users to adjust.
+The other is customized losses, in which users can define loss modules and combine loss modules as they like.
+
+## Default losses
+
+For convenient usage, you can directly use default loss calculation process we set for concrete algorithms like lsgan, biggan, styleganv2 etc.
+Take `stylegan2` as example, we use R1 gradient penalty and generator path length regularization as configurable losses, and users can adjust
```
Comment (zengyh1900) [inline docs/en/advanced_guides/losses.md:10]: take xxx as an example

File: docs/en/advanced_guides/losses.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+```
+
+As shown in this part of codes, once users set the `data_info`, the loss module will receive a dictionary containing all of the necessary data and modules, which is provided by the `MODELS` in the training procedure. If this dictionary is given as a non-keyword argument, it should be offered as the first argument. If you are using a keyword argument, please name it as `outputs_dict`.
+
+To build the computational graph, the generative models have to provide a dictionary containing all kinds of data. Having a close look at any generative model, you will find that we collect all kinds of features and modules into a dictionary. We provide a customized `GANWithCustomizedLoss` here to show the process.
+
+```python
+class GANWithCustomizedLoss(BaseModel):
+    def train_step(self, data: TrainStepInputs,
```
Comment (LeoXing1996) [inline docs/en/advanced_guides/losses.md:114]: TrainStepInputs -> dict

**Comment by LeoXing1996** [comment]
src/pytorch-sphinx-theme should not be added.

File: docs/en/advanced_guides/losses.md [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+            disc_pred_fake=disc_pred_fake,
+            disc_pred_real=disc_pred_real,
+            fake_imgs=fake_imgs,
+            real_imgs=real_imgs,
+            curr_scale=self.curr_scale[0],
+            transition_weight=transition_weight,
+            gen_partial=partial(
+                self.generator,
+                curr_scale=self.curr_scale[0],
+                transition_weight=transition_weight),
```
Comment (LeoXing1996) [inline docs/en/advanced_guides/losses.md:132]: Maybe we can remove these two lines. These attributes are designed for PGGAN.

Comment (LeoXing1996) [reply docs/en/advanced_guides/losses.md:132]: Or direct use `generator=self.generator`?

File: docs/en/advanced_guides/losses.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            disc_partial=partial(
+                self.discriminator,
+                curr_scale=self.curr_scale[0],
+                transition_weight=transition_weight))
```
Comment (LeoXing1996) [inline docs/en/advanced_guides/losses.md:136]: Same issue as `gen_partial`.

File: docs/en/advanced_guides/losses.md [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+                if loss_ is None:
+                    continue
+
+                # the `loss_name()` function return name as 'loss_xxx'
+                if loss_module.loss_name() in losses_dict:
+                    losses_dict[loss_module.loss_name(
+                    )] = losses_dict[loss_module.loss_name()] + loss_
+                else:
+                    losses_dict[loss_module.loss_name()] = loss_
+        loss, log_var = self._parse_losses(losses_dict)
```
Comment (LeoXing1996) [inline docs/en/advanced_guides/losses.md:167]: `self._parse_losses` -> `self.parse_losses` ?

File: docs/en/advanced_guides/losses.md [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+        return 'loss_disc_shift'
+
+```
+
+As shown in this part of codes, once users set the `data_info`, the loss module will receive a dictionary containing all of the necessary data and modules, which is provided by the `MODELS` in the training procedure. If this dictionary is given as a non-keyword argument, it should be offered as the first argument. If you are using a keyword argument, please name it as `outputs_dict`.
+
+To build the computational graph, the generative models have to provide a dictionary containing all kinds of data. Having a close look at any generative model, you will find that we collect all kinds of features and modules into a dictionary. We provide a customized `GANWithCustomizedLoss` here to show the process.
+
+```python
+class GANWithCustomizedLoss(BaseModel):
```
Comment (LeoXing1996) [inline docs/en/advanced_guides/losses.md:113]: We should add following code to build loss module:
```python
def __init__(self, gan_loss, disc_auxiliary_loss, gen_auxiliary_loss, *args, **kwargs):
        # ...
        if gan_loss is not None:
            self.gan_loss = MODULES.build(gan_loss)
        else:
            self.gan_loss = None

        if disc_auxiliary_loss:
            self.disc_auxiliary_losses = MODULES.build(disc_auxiliary_loss)
            if not isinstance(self.disc_auxiliary_losses, nn.ModuleList):
                self.disc_auxiliary_losses = nn.ModuleList(
                    [self.disc_auxiliary_losses])
        else:
            self.disc_auxiliary_loss = None

        if gen_auxiliary_loss:
            self.gen_auxiliary_losses = MODULES.build(gen_auxiliary_loss)
            if not isinstance(self.gen_auxiliary_losses, nn.ModuleList):
                self.gen_auxiliary_losses = nn.ModuleList(
                    [self.gen_auxiliary_losses])
        else:
            self.gen_auxiliary_losses = None
```
